### ISSUE TITLE
Problem with OVH plugin for creating/renewing certificates

### ISSUE THREAD

**Issue by mirkoglisenti** [rmbolger/Posh-ACME#537]
When you start the creation (with New-PACertificate) or you submit the renewal of a existing certificate you get an error during the Submit-ChallengeValidation phase.

Error:

Submit-ChallengeValidation: C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.20.0\Public\New-PACertificate.ps1:253Line |
 253 |          Submit-ChallengeValidation
     |          ~~~~~~~~~~~~~~~~~~~~~~~~~~
     |  {   "message": "Query out of time",   "httpCode": "400 Bad Request",   "errorCode": "QUERY_TIME_OUT" }

**Comment by mirkoglisenti** [comment]
```
New-PACertificate -Domain *.blnservice.it, *.portal.blnservice.it, blnservice.it -Plugin OVH, OVH, OVH -PluginArgs $pArgs -Verbose

DETTAGLIATO: Updating directory info from https://acme-v02.api.letsencrypt.org/directory
DETTAGLIATO: Using ACME Server https://acme-v02.api.letsencrypt.org/directory
DETTAGLIATO: Using account 381216470
DETTAGLIATO: Order name not specified, using '!.blnservice.it'
DETTAGLIATO: Creating a new order '!.blnservice.it' for *.blnservice.it, *.portal.blnservice.it, blnservice.it
DETTAGLIATO: Publishing challenge for Domain blnservice.it with Token qwkZKWmobOnl2gkl6piwwShfwLF6_zo56s8Dbssbn3I using Plugin
 OVH and DnsAlias ''.
DETTAGLIATO: GET https://eu.api.ovh.com/1.0/domain/zone/_acme-challenge.blnservice.it/record?fieldType=TXT with 0-byte payload

Submit-ChallengeValidation : {"message":"Query out of time","httpCode":"400 Bad Request","errorCode":"QUERY_TIME_OUT"}
In C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.20.0\Public\New-PACertificate.ps1:253 car:9
+         Submit-ChallengeValidation
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Submit-ChallengeValidation], Web 
   Exception
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Submit-ChallengeValidation
```

**Comment by rmbolger** [comment]
This makes it look like your machine can't reach the OVH API server for some reason. What happens if you try to just make an unauthenticated query directly to the API like this?

```powershell
Invoke-RestMethod https://eu.api.ovh.com/1.0/domain/zone/_acme-challenge.blnservice.it/record?fieldType=TXT
```

**Comment by mirkoglisenti** [comment]
`Invoke-RestMethod https://eu.api.ovh.com/1.0/domain/zone/_acme-challenge.blnservice.it/record?fieldType=TXT -verbose`
VERBOSE: Requested HTTP/1.1 GET with 0-byte payload
VERBOSE: Received HTTP/1.1 34-byte response of content type application/json
Invoke-RestMethod:
{
  "message": "You must login first"
}

**Comment by rmbolger** [comment]
Sorry for the delay on my responses. Been busy lately.

The plugin is making a query for the `_acme-challenge.blnservice.it` zone before it checks for `blnservice.it` just in case one exists. Historically, the call would result in either a 403 or 404 response if it didn't exist or the credentials hadn't been given access to it. I wonder if something in the API changed recently though.

If you're comfortable temporarily modifying the plugin file, you could tweak it so it checks for 400 instead of 403 just to see whether that is indeed the problem. It's on line 442 of the OVH.ps1 file in the Plugins folder. Just literally change 403 to 400, save the file, and force re-import the plugin.

https://github.com/rmbolger/Posh-ACME/blob/fb403a7126944c945c5d83032336dd937b8779b7/Posh-ACME/Plugins/OVH.ps1#L440-L448

You can also test just the plugin rather than a whole cert run using Publish-Challenge directly like this.

```powershell
Publish-Challenge blnservice.it (Get-PAAccount) faketoke OVH $pArgs -Verbose
```

It might be a bit before I can test this myself.

**Comment by mirkoglisenti** [comment]
Hi Ryan, I think that I've found the real problem.

After a bit of research on the "400 - query time out" error and some education about how OVH wants the query and especially the query headers to be formatted, I discovered that the problem was in the very time that was used as the timestamp to sign the http request: it was a time ahead in time (as if my server was a few seconds in the future).

After solving the time problem via windows w32t commands, I was able to fix the problem and now everything works fine.

So no problem in the Posh-ACME source code, it was a problem with my server and its time.

If the same error happens to others, I hope this helps.

Thank you very much

Translated with DeepL.com (free version)
